File: src/services/geminiService.ts
/**
 * @license
 * SPDX-License-Identifier: Apache-2.0
 */

import { GoogleGenAI, Type } from "@google/genai";

// Use the API key from process.env as per gemini-api skill
const ai = new GoogleGenAI({ apiKey: process.env.GEMINI_API_KEY || '' });

export const geminiService = {
  async summarize(content: string) {
    const response = await ai.models.generateContent({
      model: "gemini-3-flash-preview",
      contents: `Summarize this journal entry in 2-3 meaningful lines:\n\n${content}`,
      config: {
        systemInstruction: "You are a warm and supportive journaling assistant. Provide concise and meaningful summaries.",
      },
    });
    return response.text;
  },

  async detectMood(content: string) {
    const response = await ai.models.generateContent({
      model: "gemini-3-flash-preview",
      contents: `Analyze the emotional tone of this journal. Return mood, short explanation, and a gentle suggestion in JSON format.\n\n${content}`,
      config: {
        systemInstruction: "You are a warm and supportive journaling assistant. Analyze emotional tone accurately.",
        responseMimeType: "application/json",
        responseSchema: {
          type: Type.OBJECT,
          properties: {
            mood: { type: Type.STRING, enum: ['peaceful', 'reflective', 'excited', 'tense', 'gloomy', 'neutral'] },
            explanation: { type: Type.STRING },
            suggestion: { type: Type.STRING },
          },
          required: ["mood", "explanation", "suggestion"]
        }
      },
    });
    
    try {
      return JSON.parse(response.text || '{}');
    } catch (e) {
      console.error("Failed to parse mood analysis:", e);
      return null;
    }
  },

  async continueWriting(content: string) {
    const response = await ai.models.generateContent({
      model: "gemini-3-flash-preview",
      contents: `Continue or improve the user's writing naturally without sounding robotic. Here is the current text:\n\n${content}`,
      config: {
        systemInstruction: "You are a warm and supportive journaling assistant. Continue the user's writing in a matching tone.",
      },
    });
    return response.text;
  }
};

File: android/app/src/main/java/com/ethos/journal/service/GeminiService.kt
package com.ethos.journal.service

import com.google.ai.client.generativeai.GenerativeModel
import com.google.ai.client.generativeai.type.content
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject

class GeminiService(apiKey: String) {
    private val model = GenerativeModel(
        modelName = "gemini-1.5-flash",
        apiKey = apiKey
    )

    suspend fun summarize(content: String): String? = withContext(Dispatchers.IO) {
        val prompt = "Summarize this journal entry in 2-3 meaningful lines:\n\n$content"
        try {
            val response = model.generateContent(prompt)
            response.text
        } catch (e: Exception) {
            null
        }
    }

    suspend fun detectMood(content: String): MoodAnalysis? = withContext(Dispatchers.IO) {
        val prompt = "Analyze the emotional tone. Return JSON: { \"mood\": \"peaceful|reflective|excited|tense|gloomy|neutral\", \"explanation\": \"...\", \"suggestion\": \"...\" }\n\n$content"
        try {
            val response = model.generateContent(prompt)
            val json = JSONObject(response.text ?: "{}")
            MoodAnalysis(
                mood = json.optString("mood", "neutral"),
                explanation = json.optString("explanation", ""),
                suggestion = json.optString("suggestion", "")
            )
        } catch (e: Exception) {
            null
        }
    }

    suspend fun continueWriting(content: String): String? = withContext(Dispatchers.IO) {
        val prompt = "Continue this journal entry naturally:\n\n$content"
        try {
            val response = model.generateContent(prompt)
            response.text
        } catch (e: Exception) {
            null
        }
    }
}

data class MoodAnalysis(
    val mood: String,
    val explanation: String,
    val suggestion: String
)
